### PR TITLE
Add tests for some corner cases

### PR DESCRIPTION
**Pull request by djmitche**
I'll be making a list here as I think of them..

 * [x] division applies floor (or array indexes don't require ints)
 * [x] string interpolation results in a boolean
 * [x] string interpolation results in null
 * [x] string interpolation escaping
 * [x] operator escaping (how do I write a literal `{$eval: foo}`)?
 * [x] object access by named property: `{$eval: "settings[environment]"}`
 * [x] for `$if`, what values are truthy/falsy (empty strings, 0, [], {}, null)
 * [x] access of missing property in expression (`"x.foo"` when `x` has no `foo`)
 * [x] missing `then` in `$if`-`else` expression
 * [x] top-level deletion marker (`{$if: 'false', then: 1}`)
 * [x] deletion marker in an array (`[1, 2, {$if: 'false', then: 1}]`)
 * [x] `$map` on non-array things other than an object
 * [x] `$sort` over multi-digit numbers (1 < 10)
 * [x] `$sort` of mixed numbers and strings and other objects
 * [x] `$sort` of mixed numbers and strings
 * [x] `$sort` of all non-array types
 * [x] `$sort` when `by(var)` returns various types
 * [x] `$reverse` when given a non-array
 * [x] expressions: `==` does deep equality, `!=` does the opposite
 * [x] `$eval` given things other than array, string
 * [x] Numeric literals - floats, different bases, leading zeroes
 * [x] parenthetical grouping
 * [x] all parser operators (infix and unary) over all types
 * [x] string literals with `"` and with `'`
 * [x] array indexing / slicing with invalid types
 * [x] object property access with invalid types
 * [x] min/max with zero args
 * [x] all builtins with funny types
 * [x] `str([..])` formatting
 * [x] boolean literals (true, false)

### CODE REVIEW

**Comment by djmitche** [comment]
OK, this is done:

```
  763 passing (828ms)
  37 pending
```

yeah, that's a lot, and yeah, many of them are super-boring, but they do test a lot of corner cases which might otherwise be different between different implementations of JSON-e.

Where I've found problems, I've added a `todo` pointing to the github issue to fix it.  So we can remove those "todo" items in the process of fixing the issue.

**Comment by jonasfj** [comment]
More than 2k additions... I'll read that Monday :)

**Comment by djmitche** [comment]
It's probably easiest to review commit-by-commit.

**Comment by TheNavigat** [comment]
abided shows a read underline in Chrome. Abode is correct, https://www.merriam-webster.com/dictionary/abide. Apparently abided is also correct anyway :)

File: specification.yml [hunk 1]
```diff
@@ -1371,6 +1371,21 @@ context: {key: {a: 1}}
...
+---
+title: 'missing property'
+context: {key: {a: 1}}
+template: {$eval: 'key.b'}
+error: true
+---
+title: 'missing property by name'
+context: {key: {a: 1}}
+template: {$eval: 'key["b"]'}
+error: true
```
Comment (jonasfj) [inline specification.yml]: I'm not sure about this one... returning `null` or `deleteMarker` might be more powerful

File: specification.yml [hunk 2]
```diff
@@ -269,6 +269,12 @@ title: $if->then->else, nonempty object
 context: {cond: {a: 2}, key: 'world'}
 template: {$if: 'cond', then: "t", else: "f"}
 result: t
+---
+title: $if->then evaluating to nothing at the top level
+context: {cond: false}
+template: {$if: 'cond', then: "t"}
+result: {}
```
Comment (jonasfj) [inline specification.yml]: I advocate for `null`... that way `deleteMarker` is just null, but with some slightly modified semantics in list and objects...

Comment (djmitche) [inline specification.yml]: JSON distinguishes `{"x": null}` from `{}`, so I think we should too.

File: specification.yml (hunk 1 above)
Comment (djmitche) [inline specification.yml]: That's a behavior that enables typos and surprises when data formats change. I prefer missing properties or keys to be an error.

File: specification.yml [hunk 3]
```diff
@@ -2327,6 +2327,391 @@ title: 'complex logical operation (4)'
...
+error: true
+---
+title: 'number and number'
+context: {a: 345, b: 123}
+template: {$eval: "a && b"}
+error: true
+---
+title: 'number or number'
+context: {a: 345, b: 123}
+template: {$eval: "a || b"}
```
Comment (jonasfj) [inline specification.yml]: All of this is inconsistent with the fact that all other values could be used as true in `$if`

Comment (djmitche) [inline specification.yml]: It's true, and I just updated #68 to mention that.

File: specification.yml [hunk 4]
```diff
@@ -422,6 +594,31 @@ context:  {}
...
+context:  {}
+template: {$reverse: {a: 1}}
+error: true
+---
+title:    reverse of null
+context:  {}
+template: {$reverse: null}
+error: true
+---
+title:    reverse of a string
```
Comment (TheNavigat) [inline specification.yml]: Why shouldn't we be able to reverse a string?

File: specification.yml [hunk 5]
```diff
@@ -515,6 +762,11 @@ context: {}
 template: {$eval: 'sqrt(9)'}
 result: 3.0
 ---
+title: sqrt (TypeError)
```
Comment (TheNavigat) [inline specification.yml]: I think we need to add a test for `sqrt()`

File: specification.yml [hunk 6]
```diff
@@ -555,6 +807,11 @@ context: {key: 1.0}
 template: {$eval: 'ceil(key)'}
 result: 1.0
 ---
+title: ceil (TypeError)
```
Comment (TheNavigat) [inline specification.yml]: `ceil()`

File: specification.yml [hunk 7]
```diff
@@ -595,6 +852,11 @@ context: {key: 1.0}
 template: {$eval: 'floor(key)'}
 result: 1.0
 ---
+title: floor (TypeError)
```
Comment (TheNavigat) [inline specification.yml]: `floor()`

File: specification.yml [hunk 8]
```diff
@@ -695,6 +957,11 @@ context: {}
 template: {$eval: 'abs(1.0)'}
 result: 1.0
 ---
+title: abs (TypeError)
```
Comment (TheNavigat) [inline specification.yml]: `abs()`

File: specification.yml [hunk 9]
```diff
@@ -855,6 +1122,11 @@ context: {}
 template: {$eval: 'str(true)'}
 result: 'true'
 ---
+title: str (6)
```
Comment (TheNavigat) [inline specification.yml]: `str()`

File: specification.yml [hunk 10]
```diff
@@ -1056,114 +1379,1411 @@ context: {a: 2, b: 3}
...
+title: 'multiplication of strings to strings'
+context: {a: 'a', b: 'b'}
+template: {$eval: 'a * b'}
+error: true
 ---
-title: 'and operator (1)'
-context: {tt: true, ff: false}
-template: {$eval: 'tt && tt'}
-result: true
+title: 'multiplication of strings to numbers'
```
Comment (TheNavigat) [inline specification.yml]: In a certain language (I can't recall which) that basically printed the string `b` times. Do you think that might be a better approach?

File: specification.yml [hunk 11]
```diff
@@ -1056,114 +1379,1411 @@ context: {a: 2, b: 3}
...
+title: 'multiplication of strings to objects'
+context: {a: 'a', b: {x: 1}}
+template: {$eval: 'a * b'}
+error: true
 ---
-title: 'and operator with not operator'
-context: {tt: true, ff: false}
-template: {$eval: 'tt && !ff'}
-result: true
+title: 'multiplication of strings to arrays'
```
Comment (TheNavigat) [inline specification.yml]: Either change this to division, or change the `/` to `*`

File: specification.yml [hunk 12]
```diff
@@ -1056,114 +1379,1411 @@ context: {a: 2, b: 3}
...
+context: {tt: true, ff: false}
+template: {$eval: '!ff'}
+result: true
+---
+title: 'or operator (1)'
+context: {tt: true, ff: false}
+template: {$eval: 'tt || ff'}
+result: true
+---
+title: 'or operator (2)'
```
Comment (TheNavigat) [inline specification.yml]: Redundant test, identical to `or operator (1)`

File: specification.yml (hunk 1 above)
Comment (TheNavigat) [inline specification.yml]: Is it possible to add an operator or so to allow printing `null` if the property doesn't exist? In some cases that would be very useful, especially since I don't recall that we a have a "does property exist" functionality. Mmmmm, should we add that? :)

File: specification.yml [hunk 13]
```diff
@@ -1724,6 +3481,413 @@ title: 'greater than equal (4)'
...
+---
+title: equality of object and boolean
+context: {a: {}, b: true}
+template: {$eval: 'a == b'}
+result: false
+---
+title: equality of object and object
+context: {a: {}, b: {}}
+template: {$eval: 'a == b'}
+result: false
```
Comment (TheNavigat) [inline specification.yml]: Why is this false?

File: specification.yml [hunk 14]
```diff
@@ -1724,6 +3481,413 @@ title: 'greater than equal (4)'
...
+---
+title: equality of array and boolean
+context: {a: [], b: true}
+template: {$eval: 'a == b'}
+result: false
+---
+title: equality of array and array
+context: {a: [], b: {}}
```
Comment (TheNavigat) [inline specification.yml]: `array and array` should be `array and object`

File: specification.yml [hunk 15]
```diff
@@ -1724,6 +3481,413 @@ title: 'greater than equal (4)'
...
+template: {$eval: 'a == b'}
+result: false
+---
+title: equality of array and array
+context: {a: [], b: []}
+template: {$eval: 'a == b'}
+result: false
```
Comment (TheNavigat) [inline specification.yml]: Why is this false?

File: specification.yml (hunk 2 above)
Comment (jonasfj) [inline specification.yml]: but when `else` branch is missing in `{$if: 'false', then: "t"}`, wouldn't it be more correct to resolve as `null`.
Current semantics imply that: `{$if: 'false', then: "t"}` is the same as `{$if: 'false', then: "t", else: {}}` if used as top-level, but not if used inside an array...

`result: {}` is an accident... I'm pretty sure the `{}` given here is the object we call ´_deleteMarker` in the code.. .We use to signal delete-marker. In this case there is no `else` branch, so the property or array index that the value is assigned to should be deleted. But since this is top-level that won't happy.

Hence, my argument that if `_deleteMarker` is the top-level result, the result to should be transformed to `null`.

Hence, `{$if: 'false', then: "t"}` is the same as `{$if: 'false', then: "t", else: null}` if used as top-level object, both become `null`.

If instead we have `[{$if: 'false', then: "t"}]` and `[{$if: 'false', then: "t", else: null}]`, then the results are `[]` and `[null]` respectively.


-----
Current behavior:
 * `{$if: 'false', then: "t"} => {}` // happens by accident because `_deleteMarker == {}`
 * `{$if: 'false', then: "t", else: {}} => {}`
 * `{$if: 'false', then: "t", else: null} => null`
 * `[{$if: 'false', then: "t"}] => []` // intended behavior!!!
 * `[{$if: 'false', then: "t", else: {}}] => [{}]`
 * `[{$if: 'false', then: "t", else: null}] => [null]`
I suggest:
 * `{$if: 'false', then: "t"} => null` **changed this special case**
 * `{$if: 'false', then: "t", else: {}} => {}`
 * `{$if: 'false', then: "t", else: null} => null`
 * `[{$if: 'false', then: "t"}] => []`
 * `[{$if: 'false', then: "t", else: {}}] => [{}]`
 * `[{$if: 'false', then: "t", else: null}] => [null]`

File: specification.yml (hunk 4 above)
Comment (jonasfj) [inline specification.yml]: make a function for it... the `$<keyword>` constructs are intended to operate on JSON structures... They aren't intended to be auxiliary functions...

File: specification.yml (hunk 5 above)
Comment (jonasfj) [inline specification.yml]: propose the test in a PR.

File: specification.yml (hunk 15 above)
Comment (jonasfj) [inline specification.yml]: good question, didn't we have deep equality?

File: specification.yml (hunk 1 above)
Comment (jonasfj) [inline specification.yml]: > That's a behavior that enables typos and surprises when data formats change. I prefer missing properties or keys to be an error.

Any thoughts for a syntax for taking a property or falling back to a default...
The reason I like null is that often you'll want to do `key['property'] || 'default-value'`

Comment (djmitche) [inline specification.yml]: Hm, it's true, since we don't have conditional expressions, we have no way to express that (note that `||` currently requires booleans).

So I can see a few options:
 * `x['foo']` and `x.foo` are `null` - but this will not "go away" when users want it to
 * `x['foo']` and `x.foo` are a delete marker - this will require handling delete markers within the expression language, where currently they're only handled in array/object construction
 * `x['foo']` and `x.foo` are errors, but `get(x, 'foo', 123)` is `123`, and the `get` built-in always requires that third argument.
Preferences?

File: specification.yml (hunk 2 above)
Comment (djmitche) [inline specification.yml]: Ah, nice catch!

I think I intended this case to be an error -- it will be unusual to use an `$if` at the top level, and a read of a one-armed, top-level `$if` leaves the other arm undefined.  If you have a reason for that not to be an error, I don't mind null, but I'd prefer an error.  The other cases in your list above I agree with.

File: specification.yml (hunk 5 above)
Comment (djmitche) [inline specification.yml]: I can add one here :)

File: specification.yml (hunk 6 above)
Comment (djmitche) [inline specification.yml]: I don't understand this suggestion - none of the other descriptions include the empty parentheses.  Why does it matter?

File: specification.yml (hunk 10 above)
Comment (djmitche) [inline specification.yml]: Python does that.  i'm not sure it'd be useful, though.  We could add it later if it is.

File: specification.yml (hunk 15 above)
Comment (djmitche) [inline specification.yml]: Oops, yes, that's definitely a mistake.

File: specification.yml (hunk 1 above)
Comment (djmitche) [inline specification.yml]: Anyway, we can sort that out after landing this patch, and adjust the tests when we make the change.  We can do whatever we want until 1.0! :)

**Comment by djmitche** [comment]
I think I've fixed everything except the top-level conditional, since that's still an open question.  Can we merge?

**Comment by jonasfj** [comment]
I say ship it... I filed https://github.com/taskcluster/json-e/issues/77 for property access semantics...

**Comment by djmitche** [comment]
rebased on top of the Python testing patch

File: specification.yml (hunk 2 above)
Comment (TheNavigat) [inline specification.yml]: I'm not sure which is more correct.

The whole motive behind json-e is that it will always return correct JSON. While `null` is parse-able it defeats the purpose of always returning JSON, but returning `null` seems more correct than returning `{}` in that specific case.

I don't prefer an error because sometimes you might indeed have a case where you return nothing. We might force an `else` clause, but allowing `if` statements without `else` clauses and returning an error if the condition isn't met is... weird. Imagine the cases json-e can be used for, in which an if condition like that, which returns nothing in some cases, would be useful. Polling for example.

Comment (djmitche) [inline specification.yml]: Just to note, `null` is valid JSON (it represents None in Python, null in JS, NULL in Java, etc.)
